src/main/java/ch/njol/skript/expressions/ExprDyed.java

```diff
@@ -0,0 +1,128 @@
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.util.Color;
+import ch.njol.util.Kleenean;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.bukkit.inventory.meta.LeatherArmorMeta;
+import org.bukkit.inventory.meta.MapMeta;
+import org.bukkit.inventory.meta.PotionMeta;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.regex.Matcher;
+
+@Name("Dyed")
+@Description("An expression to return items/entities with a color.")
+@Examples({
+		"give player leather chestplate dyed red",
+		"give player potion of invisibility dyed rgb 200, 70, 88",
+		"give player filled map with color rgb(20, 60, 70)"
+})
+@Since("INSERT VERSION")
+public class ExprDyed extends SimpleExpression<ItemType> {
+
+	private static final boolean MAPS_AND_POTIONS_COLORS = Skript.methodExists(PotionMeta.class, "setColor", org.bukkit.Color.class);
+	
+	static {
+		Skript.registerExpression(ExprDyed.class, ItemType.class, ExpressionType.COMBINED, "%itemtypes% (dyed|with color [of]) %color%");
+	}
+	
+	private Expression<ItemType> targets;
+	private Expression<Color> color;
+
+	@Override
+	@SuppressWarnings({"unchecked", "null"})
+	public boolean init(Expression<?>[] vars, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		targets = (Expression<ItemType>) vars[0];
+		color = (Expression<Color>) vars[1];
+		return true;
+	}
+	
+	@Override
+	@Nullable
+	protected ItemType[] get(Event event) {
+		Color color = this.color.getSingle(event);
+		ItemType[] targets = this.targets.getArray(event);
+		org.bukkit.Color bukkitColor;
+
+		if (color == null)
+			return new ItemType[0];
+
+		bukkitColor = color.asBukkitColor();
+		for (ItemType item : targets) {
+			ItemMeta meta = item.getItemMeta();
+
+			if (meta instanceof LeatherArmorMeta) {
+				LeatherArmorMeta m = (LeatherArmorMeta) meta;
+				m.setColor(bukkitColor);
+				item.setItemMeta(m);
+			} else if (meta instanceof MapMeta && MAPS_AND_POTIONS_COLORS) {
+				MapMeta mapMeta = (MapMeta) meta;
+				mapMeta.setColor(bukkitColor);
+				item.setItemMeta(mapMeta);
+			} else if (meta instanceof PotionMeta && MAPS_AND_POTIONS_COLORS) {
+				PotionMeta potionMeta = (PotionMeta) meta;
+				potionMeta.setColor(bukkitColor);
+				item.setItemMeta(potionMeta);
+			} else {
+				Material material = item.getMaterial();
+				Matcher matcher = ExprColorOf.MATERIAL_COLORS_PATTERN.matcher(material.name());
+				if (!matcher.matches()) {
+					continue;
+				}
+				try {
+					Material newItem = Material.valueOf(material.name().replace(matcher.group(1), color.getName()));
+					item.setTo(new ItemType(newItem));
+				} catch (Exception ignored) {}
+			}
+		}
+		return targets.clone();
```

Comment:
> because this is not needed, this also removes the need to define ItemType[] targets = this.targets.getArray(event);

I had to leave that because we need to return the values of `targets`  and with your suggestion we will have to call getArray() twice